src/common/node/promise.ts

```diff
@@ -22,6 +22,17 @@ export class PromiseUtil {
         return this.retryAsyncIteration(operation, condition, maxRetries, 0, delay, failure);
     }
 
+    public reduce<T>(sources: T[]|Q.Promise<T[]>, generateAsyncOperation: (value: T) => Q.Promise<void>): Q.Promise<void> {
```

Comment:
At the moment I'm only using this with null.
The first time we need to use this and not return null, we can come back and modify it to support it (I basically didn't want to spend extra time trying to making it more generic, until we know that we do need it).